netmiko/cisco/cisco_bsp.py

```diff
@@ -0,0 +1,284 @@
+                time.sleep(1)
+                self.write_channel(bmc_username)
+                time.sleep(1 * delay_factor)
+                output = self.find_prompt()
+
+            log.debug("Check if BMC Password Prompt detected")
+            if re.search(pwd_pattern, output):
+                log.debug("BMC Password pattern detected, sending Password={}".format(bmc_pass))
+                self.write_channel(bmc_pass)
+                time.sleep(.5 * delay_factor)
+                output = self.find_prompt()
+
+                if re.search(prompt_pattern, output):
+                    log.debug("On BMC Prompt")
+                    return
+
+                if re.search(pwd_pattern, output):
+                    self.write_channel(bmc_pass)
+                    time.sleep(.5 * delay_factor)
+
+        # Last try to see if we already logged in
+        self.write_channel(self.TELNET_RETURN)
+        time.sleep(.5 * delay_factor)
+        output = self.find_prompt()
+        if re.search(prompt_pattern, output):
+            log.debug("On BMC Prompt")
+            return
+
+        raise NetMikoAuthenticationException("LAST_TRY login failed for BMC Prompt")
+
+    def set_base_prompt(self, pri_prompt_terminator='#',alt_prompt_terminator='$', delay_factor=1):
+        """Sets self.base_prompt
+
+        Used as delimiter for stripping of trailing prompt in output.
+
+        Should be set to something that is general and applies in multiple contexts. For Cisco
+        devices this will be set to router hostname (i.e. prompt without '>' or '#').
+
+        This will be set on entering user exec or privileged exec on Cisco, but not when
+        entering/exiting config mode.
+
+        :param pri_prompt_terminator: Primary trailing delimiter for identifying a device prompt
+        :type pri_prompt_terminator: str
+
+        :param alt_prompt_terminator: Alternate trailing delimiter for identifying a device prompt
+        :type alt_prompt_terminator: str
+
+        :param delay_factor: See __init__: global_delay_factor
+        :type delay_factor: int
+        """
+        prompt = self.find_prompt(delay_factor=delay_factor)
+        if not prompt[-1] in (pri_prompt_terminator, alt_prompt_terminator):
+            raise ValueError("BSP prompt not found: {0}".format(repr(prompt)))
+        # Strip off trailing terminator
+        self.base_prompt = prompt[:-1]
+        return self.base_prompt
+
+    def disable_paging(self, *args, **kwargs):
+        """Paging is disabled by default."""
+        return ""
+
+class CiscoBspSSH(CiscoBsp):
+    '''
+    CiscoBspSSH is based of CiscoBsp -- CiscoBaseConnection
+    '''
+
+    pass
+
+class CiscoBspTelnet(CiscoBsp):
+    '''
+    CiscoBspTelnet is based of CiscoBsp -- CiscoBaseConnection
+    '''
+
+    def telnet_login(self,pri_prompt_terminator=r']\s*\#',alt_prompt_terminator=r']\s*\$',username_pattern=r'login',pwd_pattern=r'assword',delay_factor=1,max_loops=20):
```

Comment:
Either we do add an exception in bmc to bsp or vise versa and catch and print expected vs current prompt for all negative scenario OR , we run a final check before coming out of the function to see we reached desired bsp  prompt or not . @shreyashah  discuss with her.

Reply:
Nishant is this required? Current code handles in this way:

```
                if 'bmc' in output:
                    self.bmc_login()        -->> will proceed if all good, will raise NetMikoAuthenticationException if login failure
                    self.bmc_to_bsp_prompt()   -->> will proceed if all good, will raise ValueError Prompt not found if failure
```

Do we need to try catch and handle it again over here? telnet_login method will directly raise exception through bmc_login/ bmc_to_bsp_prompt method calls